protocol/mempool/mempool.go

```diff
@@ -23,14 +25,18 @@ func New() *MemPool {
 
 // Insert adds a new pending tx to the pending tx pool.
 func (m *MemPool) Insert(ctx context.Context, tx *bc.Tx) error {
+	m.mu.Lock()
 	m.pool = append(m.pool, tx)
+	m.mu.Unlock()
 	return nil
 }
 
 // Dump returns all pending transactions in the pool and
 // empties the pool.
 func (m *MemPool) Dump(context.Context) ([]*bc.Tx, error) {
-	txs := m.pool[:len(m.pool):len(m.pool)]
-	m.pool = nil
+	m.mu.Lock()
+	txs := m.pool
+	m.pool = make([]*bc.Tx, 0, len(txs))
```

Comment:
It was intended as an estimate of what the next pool size might be. Might not be worth it if it's confusing.